File: problems/2095/solution.py
from typing import Optional


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        dummy = ListNode(0, head)
        prev = dummy
        while prev != slow and prev.next != slow:
            prev = prev.next

        prev.next = slow.next
        return dummy.next

File: problems/2095/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
 public:
  ListNode* deleteMiddle(ListNode* head) {
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast != nullptr && fast->next != nullptr) {
      fast = fast->next->next;
      slow = slow->next;
    }
    ListNode dummy(0);
    dummy.next = head;
    ListNode* prev = &dummy;
    while (prev != slow && prev->next != slow) {
      prev = prev->next;
    }
    prev->next = slow->next;
    return dummy.next;
  }
};